### PR TITLE
Add customer information to mail when we notify producers when required 

### LINKED ISSUE #6303: Notify Producers Email includes Customer Names when Customer Names Report is checked.
EDITED TO ONLY FOCUS ON CUSTOMER NAMES
NEW ISSUE CREATED FOR NAME OF ORDER CYCLE 

## Description

This issue is discussed in Discourse [here](https://community.openfoodnetwork.org/t/notify-producers-email-to-include-customer-details/1311)

**As an Enterprise:** I want to give my Producers access to the information they need to aid with packing and administration of orders.  I currently do this by sending the Notify Producers Email which includes a list of Products that have been ordered. Many of my Producers use this email to pack by Customer and need Customer First Name & Last Name to pack the products ordered

As a result of https://github.com/openfoodfoundation/openfoodnetwork/pull/5769 I can now change my shop preferences so that Customer Names appear in the Reports that my Producers run on the platform. 

However, when i send the Notify Producers email to these same Producers, Customer Names are not included. Many of my producers don't run their own reports, and the only report they receive about Order Cycles is this Notify Producers Email. It is consistent and expected behaviour to include Customer Names in this email as well as in Reports.

 **The Notify Producers Email:** is generated and sent to all Producers in an Order Cycle when the Enterprise logs into the Edit Order Cycle page and selects the Notify Producers Button.

Screen Shot 2020-11-04 at 3 33 17 PM

**CURRENT BEHAVIOUR**

Currently, the Notify Producers email includes a a list of products summed by VARIANT. It looks like this:

Screen Shot 2020-11-04 at 4 22 48 PM

**DESIRED BEHAVIOUR**

The email should include Customer First and Last Names in the Notify Producers email when I have checked the box in Shop Preferences to show Customer Names in Reports.
Screen Shot 2020-11-04 at 3 44 20 PM
When the Customer Names in Reports button is checked by an Enterprise in Shop Preferences, the Notify Producers email should include an additional, second set of data.

Data set 1: (as per current) is a list of products, with totals summed by VARIANT as per the current email. 
Data set 2: (when button checked) is a list of products, with totals summed by CUSTOMER **and** including the Customer's First and Last Name. 

When the Customer Names in Reports button is NOT checked, the Notify Producers email would only include the following data set:

**Data set 1:**
Product / Variant Name
Quantity
Price
Subtotal
Included Tax

When the Customer Names in Reports button IS checked, the Notify Producers email would include the above data set **as well as** the  following data set:

**Data set 2**
Product / Variant Name
Quantity
First Name
Last Name

The layout of the two data sets would reflect this example provided by the UK  - except that it would not have the Customer’s Last Name xxxxxx out. 
 
Screen Shot 2020-11-04 at 3 40 52 PM

## Acceptance Criteria & Tests

1.
2. 
3. 
4.

### PR DESCRIPTION
**Pull request by vishaldeepak**
#### What? Why?
Solves Issue - #6303 
Closes #6303

In the producer_mailer.rb file add code to accomodate for Table where line items are shown along with customer last name and first name.

#### What should we test?
Notify Producer email in Edit Order Cycle. Make sure that in the enterprise shop preferences , `Customer Names in Reports` is enabled. 

#### Release notes

Show Line items grouped by customer in Notify Producers email when flag `preferred_show_customer_names_to_suppliers` is set

Changelog Category: User facing changes 

Screenshot 2021-08-10 at 9 58 22 AM

### CODE REVIEW

**Comment by vishaldeepak** [comment]
Couple of questions 
- I have grouped line items by customer id of an order. But I have noticed that `first_name` and `last_name` are set on the shipping/billing address of the order. Should I be instead grouping on first_name and last_name instead provided in shipping address?
   - If we do decide to group by first_name. What happens if two different customers provide the same name?
- Can the same customer place two orders in same order cycle? I have considered that he can
- Do we need to sort customers in alphabtical order?

One clarified I will write the test cases

**Comment by lin-d-hop** [comment]
Ping @openfoodfoundation/train-drivers-product-owners

**Comment by RachL** [comment]
Hello @vishaldeepak from what I understand from the issue, the layout does not change, we are only adding first name and last name as the last 23 columns. So I'm not sure you need grouping by customer at all? Or maybe you need in technical terms, note here I'm answering with expected result on the UI :+1:

Yes the same customer can place 2 orders in a given order cycle.

I don't think we need to sort customer as again, the result is still excepted by line items?

I hope this helps, don't hesitate to ask further questions :)

**Comment by lin-d-hop** [comment]
@RachL Actually that's incorrect. The current email groups by product/variant. We DO need a grouping as we are disaggreating the orders by variant so that we can adding details per customer.

@vishaldeepak We'll have a chat on Slack and get back to you by the end of today.

**Comment by lin-d-hop** [comment]
Hi @vishaldeepak 

Having given this a little more thought, and thinking about consistency with the rest of the app:

- GROUP BY Order ID. 
- SORT BY Order Last Name, then Order First Name. In both cases alphabetically ascending.

This will mean that:
- If the same customer places 2 orders then these will appear in separate line item rows (consistent with the rest of the app)
- The two customers with the same name place orders, these won't be mistakenly grouped by the app 
- The list of names of customers will be grouped and listed alphabetically to complement the grouping by product/variant that already exists in the email.

So if we have:
Order 1 - Jane Doe = 3x apples, 2x oranges
Order 2 - Jack Foe = 1x apple, 1x orange
Order 3 - Jane Doe = 1x apple

This will be presented as:

Apples x3 - Jane Doe
Oranges x2 - Jane Doe
Apples x1 - Jane Doe
Apples x1 Jack Foe
Oranges x1 Jack Foe

Hopefully others agree and this makes sense. Let me know if you have more questions :-)

File: spec/mailers/producer_mailer_spec.rb [hunk 1]
```diff
@@ -128,6 +128,33 @@
     expect(mail.body.encoded).to include(p1.name)
   end
 
+  context 'when flag preferred_show_customer_names_to_suppliers is true' do
+    let(:s1) { create(:supplier_enterprise, preferred_show_customer_names_to_suppliers: true) }
+
+    it "adds customer names table" do
+      expect(s1.preferred_show_customer_names_to_suppliers).to eq(true)
```
Comment (andrewpbrett) [inline spec/mailers/producer_mailer_spec.rb:135]: This assertion (just line 135) seems unnecessary - we're setting it to be true right before this example.

File: spec/mailers/producer_mailer_spec.rb [hunk 2]
```diff
@@ -128,6 +128,33 @@
     expect(mail.body.encoded).to include(p1.name)
   end
 
+  context 'when flag preferred_show_customer_names_to_suppliers is true' do
```
Comment (andrewpbrett) [inline spec/mailers/producer_mailer_spec.rb:131]: We could add a single test for the case where it's not true, to ensure that the names are not displayed in that case.

**Review by andrewpbrett — APPROVED** [review]
This looks good, and it looks like it should close the issue. Couple minor comments about the specs.

File: spec/mailers/producer_mailer_spec.rb (hunk 2 above)
Comment (vishaldeepak) [reply spec/mailers/producer_mailer_spec.rb:131]: @andrewpbrett Made required changes and did a force push.

**Comment by RachL** [comment]
hello @vishaldeepak I've staged your PR but I don't manage to get customer names in the email. 

Here is the set up that I have made:

- A producer give a hub permissions to sell their products and add to inventory
- The hub opens a shop - I used 2 different email to place orders at that shop, ordering products from the producer 
- I first check that customers names did not appear, both in reports and email sent to producer - all good
- Then I checked the box in enterprise settings:
image
- Reports were showing customers names accordingly - all good here
- But the email remained the same.

What am I missing?

**Comment by vishaldeepak** [comment]
@RachL , can you confirm you hit `Notify Producers`, because the ticket mentioned that.

Screenshot 2021-08-29 at 8 25 14 PM

**Comment by RachL** [comment]
@vishaldeepak yes I did, sorry I forgot to mention.

**Comment by vishaldeepak** [comment]
@andrewpbrett How do I go ahead resolving this? This works perfectly fine on my local system. Could this be some kind of cache related problem?

**Comment by vishaldeepak** [comment]
@RachL I made an overview video of how it looks in my local. Maybe this will help us out

https://user-images.githubusercontent.com/9512872/131697930-57d94818-65c9-4f17-89b0-b3a5eb7fdb5d.mov


https://user-images.githubusercontent.com/9512872/131697944-54982c8e-01af-495a-87a1-7435f2b9f788.mov


https://user-images.githubusercontent.com/9512872/131697963-59f12da7-528d-40de-b50b-1441c43c3451.mov

**Comment by filipefurtad0** [comment]
Thank so much for this PR and for the videos @vishaldeepak!
I was also having troubles reproducing the behavior you describe, that was helpful :pray: 

It seems, the changes only take effect on the Email sent to producers, after new orders are placed. After doing this I could see the effect of your PR, but only on the first test case:

1)  OC is both supplied and coordinated by the same enterprise (Red Berries Hub); the **coordinator enterprise has the option enabled**:
image

Before this PR (left) we don't have the customer section present; after this PR (right) we can see the customer section, scrolling after scrolling down:

Peek 2021-09-01 17-04

2) Order Cycle coordinated buy the same enterprise but supplied by another producer (Cranberry Farm)

Here we can see that this is not being displayed (even after placing a new order):

Peek 2021-09-01 17-29

Which it should, as it is the coordinator of the OC that decides whether the names are displayed to the producer. This setting is not accessible in the UI of supplier profile enterprises:

image

So I think the PR requires that each supplier has this setting set to `true`, but this is not the expected behavior. To best display this behavior - and with the help from  @apricot12 - we can set this option from the supplier enterprise to `true` by running this on the rails console:

```
irb(main):016:0> enterprise = Enterprise.find(1786)
=> #<Enterprise id: 1786, name: "Cranberry Farm", description: "short-describing the shop.", long_descript...
irb(main):017:0> enterprise.set_preference(:show_customer_names_to_suppliers, true)
=> true

```

After this, triggering the notification email now displays the customers names:

image

**Summary**

This PR partially addresses the issue, but requires that each individual producer/supplier has the "display customer names" setting set to `true`, which might be feasible for shops but not for supplier only profiles. This means, that ideally customer names should be displayed on the notification Email if the coordinator of the order cycle has this option set.

Could you make this improvement on this PR still @vishaldeepak ?
Thanks again!

**Comment by RachL** [comment]
Aaah now I understand thanks @filipefurtad0 I was testing the second scenario only. 

To give a bit of background : it must be the shop that allows the producer to access the customer names. It cannot be the reverse situation. The setting is here to prevent cases where the producer / distributor relationship is really separate (the distributor does not want the producer to be able to reach the distributor's customers).

**Comment by vishaldeepak** [comment]
Made Changes accordingly in a new commit. Thanks everyone

**Review by andrewpbrett — APPROVED** [review]
This looks like it addresses the issue, ready for a re-test by @filipefurtad0 👍

**Pull request merged**